sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala

```diff
@@ -278,77 +280,186 @@ class RocksDB(
+      if (changelogReader != null) {
+        changelogReader.closeIfNeeded()
+      }
+    }
+    currLineage
+  }
+
+
   /**
    * Load the given version of data in a native RocksDB instance.
    * Note that this will copy all the necessary file from DFS to local disk as needed,
    * and possibly restart the native RocksDB instance.
    */
-  def load(
+  private def loadV2(
+        version: Long,
+        stateStoreCkptId: Option[String],
+        readOnly: Boolean = false): RocksDB = {
+  // An array contains lineage information from [snapShotVersion, version] (inclusive both end)
+  var currVersionLineage: Array[LineageItem] = lineageManager.getLineage()
+  try {
+    if (loadedVersion != version || (loadedStateStoreCkptId.isEmpty ||
+        stateStoreCkptId.get != loadedStateStoreCkptId.get)) {
+      closeDB(ignoreException = false)
+
+      val (latestSnapshotVersion, latestSnapshotUniqueId) = {
+        // Special handling when version is 0.
+        // When loading the very first version (0), stateStoreCkptId does not need to be defined
+        // because there won't be 0.changelog / 0.zip file created in RocksDB under v2.
+        if (version == 0) {
+          assert(stateStoreCkptId.isEmpty, "stateStoreCkptId should be empty when version is zero")
+          (0L, None)
+        // When there is a snapshot file, it is the ground truth, we can skip
+        // reconstructing the lineage from changelog file.
+        } else if (fileManager.existsSnapshotFile(version, stateStoreCkptId)) {
+          currVersionLineage = Array(LineageItem(version, stateStoreCkptId.get))
+          lineageManager.resetLineage(currVersionLineage)
+          (version, stateStoreCkptId)
+        } else {
+          val latestSnapshotVersionsAndUniqueIds =
+            fileManager.getLatestSnapshotVersionAndUniqueId(version)
+          currVersionLineage = getLineageFromChangelogFile(version, stateStoreCkptId) :+
+            LineageItem(version, stateStoreCkptId.get)
+          lineageManager.resetLineage(currVersionLineage)
+          if (latestSnapshotVersionsAndUniqueIds.length == 0) {
+            (0L, None)
+          } else {
+            logInfo(log"Latest snapshot version and uniqueId found: " +
+              log"${MDC(LogKeys.LINEAGE, printLineage(latestSnapshotVersionsAndUniqueIds))}.")
+            getLatestSnapshotVersionAndUniqueIdFromLineage(
+              currVersionLineage, latestSnapshotVersionsAndUniqueIds)
+          }
+        }
+      }
+
+      logInfo(log"Loaded latestSnapshotVersion: ${
+        MDC(LogKeys.SNAPSHOT_VERSION, latestSnapshotVersion)}, latestSnapshotUniqueId: ${
+        MDC(LogKeys.UUID, latestSnapshotUniqueId)}")
+
+      val metadata = fileManager.loadCheckpointFromDfs(latestSnapshotVersion,
+        workingDir, rocksDBFileMapping, latestSnapshotUniqueId)
+
+      init(version, latestSnapshotVersion, metadata)
+
+      if (loadedVersion != version) {
+        replayChangelog(version, Some(currVersionLineage))
```

Comment:
I hope we make the new argument of `replayChangelog()` more consistent. `replayChangelog()` currently loads from `loadedVersion` to `endVersion`, and load delta files of checkpointID from `currVersionLineage`. However, `currVersionLineage` can have more entries older than `loadedVersion`, and might have fewer entries, if we aren't able to find snapshot or other problems. It will be better if we clean this up before calling `replayChangelog()`, and if the lineage shows up, uses the versions shown in the cleaned lineage.

Reply:
sure, added